server/src/main/java/org/opensearch/indices/IntKeyLookupStore.java

```diff
@@ -0,0 +1,164 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.indices;
+
+/**
+ * An interface for objects that hold an in-memory record of hashes of keys in the disk cache.
+ * These objects have some internal data structure which stores some transformation of added
+ * int values. The internal representations may have collisions. Example transformations include a modulo
+ * or -abs(value), or some combination.
+ */
+public interface IntKeyLookupStore {
+    /**
+     * Transforms the input value into the internal representation for this keystore
+     * and adds it to the internal data structure.
+     * @param value The value to add.
+     * @return true if the value was added, false if it wasn't added because of a
+     * collision or if it was already present.
+     */
+    boolean add(int value) throws Exception;
+
+    /**
+     * Checks if the transformation of the value is in the keystore.
+     * @param value The value to check.
+     * @return true if the value was found, false otherwise. Due to collisions, false positives are
+     * possible, but there should be no false negatives unless forceRemove() is called.
+     */
+    boolean contains(int value) throws Exception;
+
+    /**
+     * Returns the transformed version of the input value, that would be used to stored it in the keystore.
+     * This transformation should be always be the same for a given instance.
+     * @param value The value to transform.
+     * @return The transformed value.
+     */
+    int getInternalRepresentation(int value);
+
+    /**
+     * Attempts to safely remove a value from the internal structure, maintaining the property that contains(value)
+     * will never return a false negative. If removing would lead to a false negative, the value won't be removed.
+     * Classes may not implement safe removal.
+     * @param value The value to attempt to remove.
+     * @return true if the value was removed, false if it wasn't.
+     */
+    boolean remove(int value) throws Exception;
+
+    /**
+     * Check if the implementing class supports safe removals. If it doesn't, remove() will always return false.
+     * @return true if the class supports safe removals, false if it doesn't.
+     */
+    boolean supportsRemoval();
```

Comment:
Why do we need this in the first place? We ideally should not expose any implementation which can't support removals as it is a basic/mandatory requirement.

Reply:
I was thinking that because of collisions either in the hash function or the modulo, removing values might also remove the keys which collided with those values, and we don't want to have false negatives because it might mean we have to recompute the query. In the RemovableIntKeyLookupStore we have a set of values which have had collisions which lets us safely remove un-collided values, which should be most of them. I also added a forceRemove function so you can get rid of any value for either class, at the cost of no longer guaranteeing no false negatives.